### PR TITLE
Feat/Emily can populate chainstate table with the height and hash inferred from deposit and withdrawal updates.

### PR DESCRIPTION
**Pull request by AshtonStephens**
## Description

Changes Emily to accept deposit / withdrawal updates that have unknown chainstate entries, and updates the chainstate
entries to match the chainstates that can be inferred from the deposit / withdrawal updates.

Closes: #?

## Changes

- Deposit update flow will update chainstate
- Withdrawal update flow will update chainstate

## Testing Information

Tested with integration tests.

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: emily/handler/src/database/entries/deposit.rs [hunk 1]
```diff
@@ -411,16 +418,48 @@ impl TryFrom<UpdateDepositsRequestBody> for ValidatedUpdateDepositsRequest {
...
-        let deposits = update_request
+        let mut deposits: Vec<_> = update_request
             .deposits
             .into_iter()
             .map(|i| i.try_into())
             .collect::<Result<_, Error>>()?;
+
+        // Order the updates by order of when they occur so that it's
+        // as though we got them in chronological order.
+        deposits.sort_by_key(|update: &ValidatedDepositUpdate| update.event.stacks_block_height);
```
Comment (matteojug) [inline emily/handler/src/database/entries/deposit.rs:429]: Will this mess up the order of the returned deposits in `UpdateDepositsResponse`? wrt the order in the client request.

**Review by matteojug — COMMENTED** [review]
> Changes Emily to accept deposit / withdrawal updates that have unknown chainstate entries

Was this change implemented? IIUC the request body is still expecting `last_update_height`/`last_update_block_hash`, and so far we were already able to send updates to emily (without sending chainstate updates), so I'm not exactly sure I got this point.

File: emily/handler/src/database/entries/deposit.rs (hunk 1 above)
Comment (AshtonStephens) [reply emily/handler/src/database/entries/deposit.rs:429]: I actually assumed it wouldn't matter, and that it wouldn't need to be in the same order, but if we care about that I can add that change

Comment (AshtonStephens) [reply emily/handler/src/database/entries/deposit.rs]: Fixed ~

File: emily/handler/src/database/entries/deposit.rs [hunk 2]
```diff
@@ -279,17 +281,21 @@ impl DepositEvent {
     pub fn ensure_following_event_is_valid(&self, next_event: &DepositEvent) -> Result<(), Error> {
         // Determine if event is valid.
         if self.stacks_block_height > next_event.stacks_block_height {
-            return Err(Error::InconsistentState(Inconsistency::ItemUpdate(
-                "Attempting to update a deposit with a block height earlier than it should be."
-                    .into(),
-            )));
+            let err_msg = format!(
+                "Attempting to update a deposit with a block height earlier than it should be..\n
```
[the remark this answers is not in the document]
Comment (AshtonStephens) [reply emily/handler/src/database/entries/deposit.rs]: Done.

**Comment by AshtonStephens** [comment]
> Was this change implemented? IIUC the request body is still expecting last_update_height/last_update_block_hash, and so far we were already able to send updates to emily (without sending chainstate updates), so I'm not exactly sure I got this point.

Oof, thought I responded to this one - the Emily API still expects these fields. This PR title misleading. In the past the Emily API required prior knowledge of the associated block by height and hash, but with this change it will infer the chainstates it should know about by taking them from the deposits and withdrawals. This means it no longer requires prior knowledge of the blocks, but it still requires the block data.

File: emily/handler/tests/integration/deposit.rs [hunk 3]
```diff
@@ -493,13 +496,23 @@ async fn update_deposits_updates_chainstate() {
...
             last_update_block_hash: format!("hash_{}", update_block_height),
-            last_update_height: update_block_height,
+            last_update_height: update_block_height as u64,
             status: update_status.clone(),
             status_message: update_status_message.into(),
         };
         deposit_updates.push(deposit_update);
     }
 
+    // Order the, pecularily so that they are not in order.
```
Comment (matteojug) [inline emily/handler/tests/integration/deposit.rs:506]: nit: missing word, also in the withdrawal test.

**Review by matteojug — APPROVED** [review]

**Pull request merged**